### PR TITLE
Generic Enumerable.Range method

### LINKED ISSUE #97689: [API Proposal]: Generic Enumerable.Range method
### Background and motivation

So that Enumerable.Range can be used with anything that implements `IBinaryInteger<T>`.

### API Proposal

```csharp
public static class Enumerable
{
    public static IEnumerable<T> Range<T>(T start, T end) where T : IBinaryInteger<T>
}
```

### API Usage

```csharp
public record struct MyInterval<T> where T : IBinaryInteger<T>
{
    public T Start { get; }
    public T End { get; }

    // ...
    
    public IEnumerable<T> GetAll() => Enumerable.Range(Start, End);
```

### Alternative Designs

_No response_

### Risks

_No response_

### PR DESCRIPTION
**Pull request by AlexRadch**
Close #97689

### CODE REVIEW

File: src/libraries/System.Linq/src/System/Linq/Range.cs [hunk 1]
```diff
@@ -24,35 +24,63 @@ public static IEnumerable<int> Range(int start, int count)
                 return [];
             }
 
-            return new RangeIterator(start, count);
+            return new RangeIterator<int>(start, count);
+        }
+
+        public static IEnumerable<T> Range<T>(T start, int count) where T : IBinaryInteger<T>
+        {
+            if (count < 0)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Linq/src/System/Linq/Range.cs:32]: Curious why this overload is using somewhat different validation flow compared to the existing method? Is it anticipating that `T.CreateTruncating` might be expensive?

File: src/libraries/System.Linq/tests/RangeIBinaryIntegerTests.cs [hunk 2]
```diff
@@ -0,0 +1,1089 @@
...
+{
+    public abstract class RangeIBinaryIntegerTests<T> : EnumerableTests where T : IBinaryInteger<T>
+    {
+#pragma warning disable xUnit1015
+
+        //public static TheoryData<T, int> StartCountCorrectData;
+        //public static TheoryData<T, int> StartCountIncorrectData;
+
+        [Theory]
+        [MemberData("StartCountCorrectData")]
```
Comment (eiriktsarpalis) [inline src/libraries/System.Linq/tests/RangeIBinaryIntegerTests.cs:18]: Can we use `nameof` here?

File: src/libraries/System.Linq/src/System/Linq/Range.cs [hunk 3]
```diff
@@ -73,27 +101,33 @@ public override void Dispose()
             {
                 _state = -1; // Don't reset current
             }
+
+            internal static int StartMaxCount(T start, T max)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Linq/src/System/Linq/Range.cs:105]: I'm guessing this is effectively trying to measure a distance between `start` and `max`? I'm not sure if this is necessarily a well-defined operation for every kind of binary integer type (e.g. Gaussian integers)

cc @tannergooding

File: src/libraries/System.Linq/tests/RangeIBinaryIntegerTests.cs [hunk 4]
```diff
@@ -0,0 +1,1089 @@
...
+            { long.MaxValue - 1, int.MinValue },
+
+            { long.MaxValue, 2 },
+            { long.MaxValue, int.MaxValue },
+            { long.MaxValue, -1 },
+            { long.MaxValue, int.MinValue },
+        };
+    }
+
+    public class RangeBigIntegerTests : RangeIBinaryIntegerTests<BigInteger>
```
Comment (eiriktsarpalis) [inline src/libraries/System.Linq/tests/RangeIBinaryIntegerTests.cs:1052]: It would be interesting to see a few nonstandard numeric types that stress test the implementation, e.g. a modular arithmetic type. If for example I define a type representing all numbers modulo 7, what should be the behaviour of the new method when I do `Range(Modulo7.Zero, 8)`?

**Review by eiriktsarpalis — COMMENTED** [review]
The build errors are caused by an [expected source breaking change](https://github.com/dotnet/runtime/issues/97689#issuecomment-2308916811) that this new API is introducing. The impacted call sites should be updated with an explicit `int` type parameter.

**Review by eiriktsarpalis — CHANGES_REQUESTED** [review]
Marking as changes requested since there are a few unresolved questions on a number of corner cases.

File: src/libraries/Common/tests/System/IO/VirtualDriveHelper.Windows.cs [hunk 5]
```diff
@@ -87,7 +87,7 @@ char GetNextAvailableDriveLetter()
                 List<char> existingDrives = DriveInfo.GetDrives().Select(x => x.Name[0]).ToList();
 
                 // A,B are reserved, C is usually reserved
-                IEnumerable<int> range = Enumerable.Range('D', 'Z' - 'D');
+                IEnumerable<int> range = Enumerable.Range((int)'D', 'Z' - 'D');
```
Comment (jkotas) [inline src/libraries/Common/tests/System/IO/VirtualDriveHelper.Windows.cs:90]:
```suggestion
                IEnumerable<int> range = Enumerable.Range<int>('D', 'Z' - 'D');
```

Comment (jkotas) [reply src/libraries/Common/tests/System/IO/VirtualDriveHelper.Windows.cs:90]: Similar in other places. Explicit selecting the result type and leave it to the compiler to convert the arguments to it implicitly is less bug prone.

**Comment by stephentoub** [comment]
Now that https://github.com/dotnet/runtime/pull/111685 has been merged, we should augment this PR to add the corresponding generic Range method to AsyncEnumerable as well. In general, we want to keep the APIs consistent across Enumerable and AsyncEnumerable.

**Comment by AlexRadch** [comment]
> Now that #111685 has been merged, we should augment this PR to add the corresponding generic Range method to AsyncEnumerable as well. In general, we want to keep the APIs consistent across Enumerable and AsyncEnumerable.

@stephentoub Done!

File: src/libraries/System.Linq.AsyncEnumerable/ref/System.Linq.AsyncEnumerable.cs [hunk 6]
```diff
@@ -110,6 +110,9 @@ public static partial class AsyncEnumerable
         public static System.Linq.IOrderedAsyncEnumerable<T> Order<T>(this System.Collections.Generic.IAsyncEnumerable<T> source, System.Collections.Generic.IComparer<T>? comparer = null) { throw null; }
         public static System.Collections.Generic.IAsyncEnumerable<TSource> Prepend<TSource>(this System.Collections.Generic.IAsyncEnumerable<TSource> source, TSource element) { throw null; }
         public static System.Collections.Generic.IAsyncEnumerable<int> Range(int start, int count) { throw null; }
+#if NET7_0_OR_GREATER
+        public static System.Collections.Generic.IAsyncEnumerable<T> Range<T>(T start, int count) where T : System.Numerics.IBinaryInteger<T> { throw null; }
+#endif
```
Comment (stephentoub) [inline src/libraries/System.Linq.AsyncEnumerable/ref/System.Linq.AsyncEnumerable.cs:115]: Thanks. This should be in its own System.Linq.AsyncEnumerable.netcore.cs file rather than ifdef'd here.

File: src/libraries/System.Linq.AsyncEnumerable/src/System/Linq/Range.cs [hunk 7]
```diff
@@ -35,5 +36,51 @@ static async IAsyncEnumerable<int> Impl(int start, int count)
...
+            }
+
+            return Impl(start, count);
+
+            static async IAsyncEnumerable<T> Impl(T start, int count)
+            {
+                for (int i = 0; i < count; i++, start++)
+                {
+                    yield return start;
+                }
```
Comment (stephentoub) [inline src/libraries/System.Linq.AsyncEnumerable/src/System/Linq/Range.cs:73]: @tannergooding, is this the right implementation? The above non-generic does:
```C#
for (int i = 0; i < count; i++)
{
    yield return start + i;
}
```
Is this proposed implementation better or worse than:
```C#
for (int i = 0; i < count; i++)
{
    yield return start + T.CreateTruncating(i);
}
```
?  I'm not sure what requirements we place on IBinaryInteger, in terms of guarantees we make about ++ actually incrementing the value (as opposed to, say, a float + 1 which might not actually increment).

File: src/libraries/System.Linq.AsyncEnumerable/src/System/Linq/Range.cs [hunk 8]
```diff
@@ -35,5 +36,51 @@ static async IAsyncEnumerable<int> Impl(int start, int count)
...
+            }
+
+            static int StartMaxCount(T start, T max)
+            {
+                int count = int.CreateTruncating(max - start);
+                if (count < 0)
+                    count = int.CreateTruncating(max) - int.CreateTruncating(start);
+                return count;
+            }
```
Comment (stephentoub) [inline src/libraries/System.Linq.AsyncEnumerable/src/System/Linq/Range.cs:82]: @tannergooding, is this how we'd recommend doing overflow checking?

**Comment by stephentoub** [comment]
@AlexRadch, thanks for working on this. It's really challenging to implement the agreed upon signature (`T start, int count`) efficiently and correctly (this PR doesn't fully account for overflow, and it's not clear there's a great way to do so). I'm going to close this PR for now and mark the API as being up for review again so that we can talk about alternatives.
